### PR TITLE
Add welltest_extract script and corresponding tests

### PR DESCRIPTION
**Pull request by eivindsm**

### CODE REVIEW

**Comment by ertomatic** [comment]
Can one of the admins verify this patch?

File: src/subscript/welltest_extract/welltest_extract.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+  -wwpr:well_name,
+  -wopr:well_name if main_phase == OIL
+  -wgpr:well_name if main_phase == GAS
+"""
+
+"""
+
+Script is a rewrite of a legacy script originally developed and improved by:
+ * Jon Saetrom
+ * Bjoern Kaare Hegstad
```
Comment (berland) [inline src/subscript/welltest_extract/welltest_extract.py:25]: utf-8 works fine! ;)

File: src/subscript/welltest_extract/welltest_extract.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        type=str,
+        help="Prefix to be added to result files, can contain a path",
+    )
+    parser.add_argument(
+        "-n",
+        "--buildup_nr",
+        type=int,
+        help="Buildup number, indicating which build up to extract. Counting from 1 ",
+        default=1,
+        required=False,
```
Comment (berland) [inline src/subscript/welltest_extract/welltest_extract.py:73]: An option is by default not required, so `required=False` can be dropped.

File: src/subscript/welltest_extract/welltest_extract.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+
+def get_weighted_avg_press_time_derivative_lag1(dp, dspt):
+    """
+    Compute weighted average of pressure time derivative,
+    one time step to each side. Lag1
+
+    Formula: (  (dp_f/dspt_f)*dspt_b + (dp_b/dspt_b)*dspt_f )/(dspt_f + dspt_b)
+
+    spt is SuperPositionedTime and dspt is delta spt
+    """
```
Comment (berland) [inline src/subscript/welltest_extract/welltest_extract.py:164]: For later refactoring, we need to know the types of the function arguments. Add to docstring?

File: src/subscript/welltest_extract/welltest_extract.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+    if buildup_nr > len(buildup_indices):
+        sys.stderr.write(
+            "There are %d build ups detected for well %s. You asked for nr %d\n"
+            % (len(buildup_indices), well_name, buildup_nr)
+        )
+        sys.exit(1)
+    else:
+        bu_start_ind = buildup_indices[buildup_nr - 1]
+        print(
+            "The time step for the start of the %d'th build-up period is %d"
```
Comment (berland) [inline src/subscript/welltest_extract/welltest_extract.py:303]: "build up", "buildup" or "build-up". We should decide.. Suggest "buildup"

**Comment by berland** [comment]
There is some refactoring that can be done in the script, but some of it can be saved for later. What we need to have firm in the first revision is the "api", that is the input arguments and the output. Input (command line arguments) look ok, but we should consider if dumping multiple CSV files is a good idea. Does it make sense to output only one? 

What consumers do we have for the CSV files? Webviz? Spotfire, user scripts? Excel?  (i.e. how  hard is it to change the output)

File: src/subscript/welltest_extract/welltest_extract.py (hunk 1 above)
Comment (eivindsm) [reply src/subscript/welltest_extract/welltest_extract.py]: Old habits, thanks!

File: src/subscript/welltest_extract/welltest_extract.py (hunk 2 above)
Comment (eivindsm) [reply src/subscript/welltest_extract/welltest_extract.py]: Fixed

File: src/subscript/welltest_extract/welltest_extract.py (hunk 3 above)
Comment (eivindsm) [reply src/subscript/welltest_extract/welltest_extract.py:193]: Agree, docstrings updated all over

File: src/subscript/welltest_extract/welltest_extract.py (hunk 4 above)
Comment (eivindsm) [reply src/subscript/welltest_extract/welltest_extract.py]: Fixed!

With regards the multiple csv files; this has been nagging me as well, but I have not been able to find a good solution. The files with one column are typically used as input to ERTs GEN_DATA, and are not csv files as such. The files with two columns, are used for visualization in eg Spotfire I guess. These could probably be merged into one unified file. We are still left with at least 4 files, and I don't like it. Another issue is wwpr, which is parsed and exported to file, but never used by the script. I would have preferred not having it as a requirement. An option would be to complexify the cli options and let the user configure.

File: tests/test_welltest_extract.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+    dpdspt = welltest_extract.get_weighted_avg_press_time_derivative_lag1(d_press, dspt)
+
+    assert len(dpdspt) == 247
+    assert dpdspt[0] == pytest.approx(0.46972867)
+    assert dpdspt[-1] == pytest.approx(0.12725929)
+
+
+def test_get_weighted_avg_press_time_derivative_lag2():
+    """ Test that weighted_avg_press_time_derivative_lag2 is calcuated correctly """
```
Comment (berland) [inline tests/test_welltest_extract.py:101]: typo calcuated

**Comment by berland** [comment]
For CSV-files (or txt-files), this is similar to `gendata_rft` in semeio. There a command line option `outputdirectory` was added for where the file collection could be dumped. 

Is GEN_DATA able to read these csv-files?

File: tests/test_welltest_dpds.py [hunk 6]
```diff
@@ -24,42 +24,42 @@ def test_main(tmpdir):
     datafilepath = os.path.join(ECLDIR, ECLCASE)
 
     # defaults only
-    sys.argv = ["welltest_extract", datafilepath, "55_33-1"]
-    welltest_extract.main()
+    sys.argv = ["welltest_dpds", datafilepath, "55_33-1"]
```
Comment (berland) [inline tests/test_welltest_dpds.py:27]: SIB prefers not to punch sys.argv, and use  `mocker` instead. See e.g. `test_summaryplot` (this mocker is not in use in all tests yet)

Comment (berland) [reply tests/test_welltest_dpds.py:27]: And potentially cleaner test function by using pytest.mark.parametrize

File: tests/test_welltest_dpds.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    assert bu_start == []
+    assert bu_end == []
+
+    wopr = welltest_dpds.get_summary_vec(summary, "WOPR:55_33-1")
+    bu_start, bu_end = welltest_dpds.get_buildup_indices(wopr)
+    assert bu_start == [7, 260]
+    assert bu_end == [254, 555]
+
+
+def test_get_supertime():
```
Comment (berland) [inline tests/test_welltest_dpds.py:96]: We should have a couple of additional tests with explicit arrays as input, not requiring a summary file to test the function. Starting with the empty array, and then increasing number of elements. pytest.mark.parametrize is a good fit.

**Comment by berland** [comment]
RST file with simple docs in docs/script can be added.

File: src/subscript/config_jobs/WELLTEST_DPDS [hunk 8]
```diff
@@ -0,0 +1,14 @@
+EXECUTABLE  welltest_dpds
+
+DEFAULT  <PHASE> OIL
+DEFAULT  <OUTPUTDIRECTORY> .
+DEFAULT  <OUTFILESSUFIX> ""
```
Comment (berland) [inline src/subscript/config_jobs/WELLTEST_DPDS:5]: sufix -> suffix

File: src/subscript/welltest_dpds/welltest_dpds.py [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+Outputs the following files according to naming convention:
+  outputdirectory/key_outfilesuffix.csv
+  - dpds_lag1 eg superpositioned time derivative of pressure lag 1
+  - dpds_lag2 eg superpositioned time derivative of pressure lag 2
+  - sspt superpositioned time
+  - wbhp vs cum time
+  - wwpr vs cum time
+  - wopr vs cum time if --phase == OIL
+  - wgpr vs cum time if --phase == GAS
+
```
Comment (berland) [inline src/subscript/welltest_dpds/welltest_dpds.py:29]: This yields a RST warning: /home/berland/projects/subscript/src/subscript/welltest_dpds/welltest_dpds.py:docstring of subscript.welltest_dpds.welltest_dpds.DESCRIPTION:17: WARNING: Definition list ends without a blank line; unexpected unindent.

File: src/subscript/welltest_dpds/welltest_dpds.py [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+"""
+
+"""
+
+Script is a rewrite of a legacy script originally developed and improved by:
```
Comment (berland) [inline src/subscript/welltest_dpds/welltest_dpds.py:34]: Formatting in this docstring gets broken in sphinx API.

File: src/subscript/welltest_dpds/welltest_dpds.py [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+    """
+
+    try:
+        return summary.numpy_vector(key)
+    except KeyError:
+        print("No such key in summary file:" + key)
+        raise
+
+
+def get_buildup_indices(rates):
```
Comment (berland) [inline src/subscript/welltest_dpds/welltest_dpds.py:137]: The `get_` prefix to many of this functions can probably just be dropped. These are not typical "getters"

File: src/subscript/welltest_dpds/welltest_dpds.py [hunk 12]
```diff
@@ -0,0 +1,475 @@
...
+import sys
+import os
+import argparse
+import numpy as np
+from ecl.summary import EclSum
+
+
+DESCRIPTION = """
+Script to extract simulated welltest results from simulator output.
+Typically used to compare with output from a welltest analysis, eg using Cappa
```
Comment (berland) [inline src/subscript/welltest_dpds/welltest_dpds.py:12]: Cappa is spelled Kappa

File: src/subscript/welltest_dpds/welltest_dpds.py [hunk 13]
```diff
@@ -0,0 +1,475 @@
...
+  * wgpr:well_name if phase == GAS
+
+Outputs the following files:
+  * dpdspt_lag1 eg superpositioned time derivative of pressure lag 1
+  * dpdspt_lag2 eg superpositioned time derivative of pressure lag 2
+  * spt superpositioned time
+
+according to the naming convention; outputdirectory/key_outfilesuffix.csv
+
+And a uinfied csv file with the following vectors:
```
Comment (berland) [inline src/subscript/welltest_dpds/welltest_dpds.py:26]: uinfied -> unified

File: src/subscript/welltest_dpds/welltest_dpds.py [hunk 14]
```diff
@@ -0,0 +1,475 @@
...
+        "--buildup_nr",
+        type=int,
+        help="Buildup nr, eg which buildup to extract. Counting from 1. Default: 1 ",
+        default=1,
+    )
+    parser.add_argument(
+        "-o",
+        "--outputdirectory",
+        type=str,
+        help="Directory to put the output files. Detault: .",
```
Comment (berland) [inline src/subscript/welltest_dpds/welltest_dpds.py:105]: Detault -> Default

File: src/subscript/welltest_dpds/welltest_dpds.py (hunk 12 above)
Comment (eivindsm) [reply src/subscript/welltest_dpds/welltest_dpds.py:12]: Ups, from this you may infer that the author is NOT a welltesting expert :)

File: src/subscript/welltest_dpds/welltest_dpds.py (hunk 14 above)
Comment (berland) [inline src/subscript/welltest_dpds/welltest_dpds.py:105]: The default is already provided in the `default` argument. To get defaults values into `welltest_dpds --help` there is another trick

Comment (berland) [reply src/subscript/welltest_dpds/welltest_dpds.py:105]: https://github.com/equinor/subscript/blob/3f06533e0d1ce94c9ab6ab0996c8beabf794829e/src/subscript/presentvalue/presentvalue.py#L34

File: src/subscript/welltest_dpds/welltest_dpds.py [hunk 15]
```diff
@@ -0,0 +1,475 @@
...
+    for i in range(len(field_list[0][start:end])):
+        fileh.write("%0.10f" % field_list[0][i])
+        for field in field_list[1:]:
+            if len(field):
+                fileh.write(sep + "%0.10f" % field[i])
+            else:
+                fileh.write(sep)
+        fileh.write("\n")
+    fileh.close()
+    print("Writing file:" + filen)
```
Comment (berland) [inline src/subscript/welltest_dpds/welltest_dpds.py:345]: Add a space after colon.

File: src/subscript/welltest_dpds/welltest_dpds.py [hunk 16]
```diff
@@ -0,0 +1,475 @@
...
+    dpdspt_weighted_lag1 = get_weighted_avg_press_time_derivative_lag1(dp, dspt)
+
+    dpdspt_weighted_lag2 = get_weighted_avg_press_time_derivative_lag2(
+        dp, dspt, super_time, wbhp, bu_start_ind, bu_end_ind
+    )
+
+    to_csv(
+        outdir + "dpdspt_lag1" + outf_suffix + ".csv",
+        [cum_time, dpdspt_weighted_lag1],
+        ["Hours", "dpd(supt)_w"],
```
Comment (berland) [inline src/subscript/welltest_dpds/welltest_dpds.py:449]: I think I would prefer it this was written "HOUR" and not "Hours", for consistency with other tools.

File: src/subscript/welltest_dpds/welltest_dpds.py [hunk 17]
```diff
@@ -0,0 +1,475 @@
...
+       buildup_end_indices (list)
+
+    """
+
+    buildup_indices = []
+    buildup_end_indices = []
+
+    last = 0
+    for i, rate in enumerate(rates):
+        if rate == 0 and last > 0.0:
```
Comment (berland) [inline src/subscript/welltest_dpds/welltest_dpds.py:164]: `rate == 0` is a floating point comparison. Might be safer to do `np.isclose(rate, 0)`

File: src/subscript/welltest_dpds/welltest_dpds.py [hunk 18]
```diff
@@ -0,0 +1,475 @@
...
+        if required:
+            raise
+    return np.array([])
+
+
+def get_buildup_indices(rates):
+    """
+    Go through the simulated rate and identify bu periods, as defined by zero flow.
+
+    Returns:
```
Comment (berland) [inline src/subscript/welltest_dpds/welltest_dpds.py:146]: Return values documented twice..

**Review by berland — APPROVED** [review]
Ok for now, remaining comments to be fixed later.

**Pull request closed without merging**